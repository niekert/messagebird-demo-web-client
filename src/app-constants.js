// Status codes
export const INITIAL = 'INITIAL';
export const PENDING = 'PENDING';
export const OK = 'OK';
export const ERROR = 'ERROR';
