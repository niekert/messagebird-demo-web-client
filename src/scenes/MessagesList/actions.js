import * as messagesApi from 'api/messages';
import formatMessage from 'shared/util/formatMessage';

export const FETCH_MESSAGES_LIST = 'FETCH_MESSAGES_LIST';
export const FETCH_MESSAGES_LIST_SUCCESS = 'FETCH_MESSAGES_LIST_SUCCESS';
export const FETCH_MESSAGES_LIST_ERROR = 'FETCH_MESSAGES_LIST_ERROR';

export const ADD_NEW_MESSAGE = 'ADD_NEW_MESSAGE';

export function fetchMessages() {
  return async dispatch => {
    dispatch({ type: FETCH_MESSAGES_LIST });

    try {
      const response = await messagesApi.fetchMessages();

      dispatch({
        type: FETCH_MESSAGES_LIST_SUCCESS,
        payload: {
          messages: response.items.map(formatMessage),
          nextUrl: response.links.next,
        },
      });
    } catch (error) {
      dispatch({ type: FETCH_MESSAGES_LIST_ERROR, payload: error });
    }
  };
}

export function addMessage(message) {
  return {
    type: ADD_NEW_MESSAGE,
    payload: {
      message,
    },
  };
}
