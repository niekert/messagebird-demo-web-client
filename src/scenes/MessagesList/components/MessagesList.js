import React from 'react';
import { arrayOf, shape } from 'prop-types';
import styled from 'styled-components';
import { status as statusPropType } from 'customPropTypes';
import { PENDING } from 'app-constants';
import { TransitionGroup } from 'react-transition-group';
import { H1 } from 'style/Headers';
import MessagesLoading from './MessagesLoading';
import Message from './Message';
import SlideIn from './SlideIn';

const Wrapper = styled.div`
  position: relative;
`;

const MessagesList = ({ messages, status }) => (
  <Wrapper>
    <H1>Messages</H1>
    {messages.length > 0 && (
      <TransitionGroup>
        {messages.map(message => (
          <SlideIn key={`message-list-${message.id}`}>
            <Message {...message} />
          </SlideIn>
        ))}
      </TransitionGroup>
    )}
    {status === PENDING && <MessagesLoading />}
  </Wrapper>
);

MessagesList.propTypes = {
  messages: arrayOf(shape(Message.propTypes)).isRequired,
  status: statusPropType.isRequired,
};

export default MessagesList;
