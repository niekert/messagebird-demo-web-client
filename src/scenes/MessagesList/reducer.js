import { createReducer } from 'redux-create-reducer';
import { INITIAL, PENDING, OK, ERROR } from 'app-constants';
import {
  FETCH_MESSAGES_LIST,
  FETCH_MESSAGES_LIST_SUCCESS,
  FETCH_MESSAGES_LIST_ERROR,
  ADD_NEW_MESSAGE,
} from './actions';

const initialState = {
  status: INITIAL,
  messages: [],
};

export default createReducer(initialState, {
  [FETCH_MESSAGES_LIST]() {},
  [FETCH_MESSAGES_LIST_SUCCESS]() {},
  [FETCH_MESSAGES_LIST_ERROR]() {},
  [ADD_NEW_MESSAGE](state, action) {
    return {
      ...state,
      messages: [action.payload.message, ...state.messages],
    };
  },
});
