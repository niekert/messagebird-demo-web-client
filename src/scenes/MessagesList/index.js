import React from 'react';
import { func } from 'prop-types';
import { connect } from 'react-redux';
import { NEW_MESSAGES_WS_URL } from 'app-constants';
import WebSocket from 'react-websocket';
import { fetchMessages, addMessage } from './actions';
import MessagesList from './components/MessagesList';

function mapStateToProps(state) {
  const { messagesList: { messages, status } } = state;

  return {
    messages,
    status,
  };
}

class MessageListContainer extends React.Component {
  static propTypes = {
    fetchMessages: func.isRequired,
    addMessage: func.isRequired,
  };

  componentDidMount() {
    this.props.fetchMessages();
  }

  onWebsocketMessage = messageJson => {
    const { type, data } = JSON.parse(messageJson);
    if (type === 'message') {
      this.props.addMessage(data);
    }

    if (type === 'connect') {
      // TODO: We could keep some state to display in the UI
      // That we're connected to the websocket
      console.log("websocket connected"); // eslint-disable-line
    }
  };

  render() {
    return [
      <MessagesList key="messageslist" {...this.props} />,
      <WebSocket
        key="new-messages-socket"
        url={NEW_MESSAGES_WS_URL}
        onMessage={this.onWebsocketMessage}
      />,
    ];
  }
}

export default connect(mapStateToProps, { fetchMessages, addMessage })(
  MessageListContainer,
);
