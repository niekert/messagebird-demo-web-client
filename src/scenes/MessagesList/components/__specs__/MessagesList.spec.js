import React from 'react';
import { OK } from 'app-constants';
import { shallow } from 'enzyme';
import MessagesList from '../MessagesList';
import formattedMessageFixture from 'fixtures/formattedMessage.json';

describe('<MessagesList />', () => {
  test('renders correctly', () => {
    const wrapper = shallow(
      <MessagesList status={OK} messages={[{ ...formattedMessageFixture }]} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
