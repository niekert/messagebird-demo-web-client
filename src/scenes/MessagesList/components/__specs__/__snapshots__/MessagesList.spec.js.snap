// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessagesList /> renders correctly 1`] = `
<MessagesList__Wrapper>
  <Headers__H1>
    Messages
  </Headers__H1>
  <TransitionGroup
    childFactory={[Function]}
    component="div"
  >
    <SlideIn
      key="message-list-a13be82534544a508c8b37ccc672eb40"
    >
      <Message
        body="Emoji support test 🐊"
        createdDatetime="2017-11-17T11:26:44+00:00"
        direction="mo"
        id="a13be82534544a508c8b37ccc672eb40"
        originator="31634322664"
        recipient={447860039373}
      />
    </SlideIn>
  </TransitionGroup>
</MessagesList__Wrapper>
`;
